### ISSUE TITLE
WORKCHAIN / DAEMON submit does not store the wc output nodes (self.out) (early finish?)

### ISSUE THREAD

**Issue by broeder-j** [aiidateam/aiida_core#622]
Executing a workchain with 'submit' will end up without the 'self.out' nodes being stored (in run and async they are).

(It did work in the tutorial version in May, or was everything just 'run'?)
AiiDA version: develop c5f6c38 (current, Django backend)
I'm not sure if this is a problem of submit or the daemon. Do you also encounter this problem?

##########
from aiida.work.run import submit
res = fleur_scf_wc.run(wf_parameters=wf_para, structure=s, inpgen = code, fleur=code2)
res = submit(fleur_scf_wc, wf_parameters=wf_para, structure=s,  inpgen = code, fleur=code2)
##########

will lead to: (verdi work list)
 4194  5m ago           fleur_scf_wc         True (submit)
 4195  5m ago           fleur_scf_wc         True (run)

verdi node show 4194:
-----------  ------------------------------------
type         WorkCalculation
pk           4194
...
##### OUTPUTS:
Link label      PK  Type
------------  ----  ------------------------
CALL          4197  FleurinputgenCalculation
CREATE        4204  ParameterData
CALL          4205  WorkCalculation
CALL          4207  FleurCalculation
##### NOTE! There are 6 log messages for this calculation.
      Use the 'calculation logshow' command to see them.

Versus verdi node show 4195 (note: nodes (fleurinp, output_scf_wc_para) from self.out are stored):
-----------  ------------------------------------
type         WorkCalculation
pk           4195
...

##### OUTPUTS:
Link label            PK  Type
------------------  ----  ------------------------
CALL                4196  FleurinputgenCalculation
CREATE              4208  ParameterData
CALL                4209  WorkCalculation
CALL                4211  FleurCalculation
fleurinp            4203  FleurinpData
output_scf_wc_para  4218  ParameterData
output_scf_wc_para  4218  ParameterData
##### NOTE! There are 6 log messages for this calculation.
      Use the 'calculation logshow' command to see them.

**Comment by broeder-j** [comment]
The same occurs kind of with different combinations of run/async and submit. as long they are executed 'two' times behind each other. (start the workchain with async and a subworkchain with async)
I am not sure that this is the final patter. just to give you something to test.

**Comment by broeder-j** [comment]
Here is a simple test system:

[
[dummy.py.txt](https://github.com/aiidateam/aiida_core/files/1127154/dummy.py.txt)
[run_dummy.py.txt](https://github.com/aiidateam/aiida_core/files/1127155/run_dummy.py.txt)
](url)

delete the .txt at the end. (could not upload a .py file).

change the import of the dummy workflow from where to import it in the run file.

async and run do work fine. running the dummy workflow with submit will work but the output node (self.out) will not be saved. submitting the sub workflow does also not work. (so far I have not used try points for the dummy workflow)

**Comment by broeder-j** [comment]
The entry points are just for the workflow factory, or are they also used for something else?

**Comment by broeder-j** [comment]
After a sphuber took a look at it. Thanks a lot! The dummy example works now. I can only say: Do not forget to run a verdi setup again for all your profiles, after you migrated your dbs... Otherwise I have now clue why this

**Comment by broeder-j** [comment]
UPDATE: After the dummy worked. My work chains unfortunately still had the same problem. But this time I knew it was on my side or a bug with something Sebastiaan does not do.

There were two things different:
1. the base qe work chain, only gives back nodes already stored
2. no reports in the last return step

The first point was ruled out by the dummy wc.
After removing my self.reports in the return step, submit also stored the output nodes. Jippi!

putting them back in resulted again in the nodes not being stored.


Overall I have the impression that there might be a clash somewhere with (several) self.report + submit (and/or database connections).
Also sometimes I have other 'strange/unpredictable changing' behavior of submit,  i.e that the work chains sometimes just start and stops after the first step with first reports (in work list they are True).
Or that it is 'started' two times, under one work chain node.

I will keep this issue closed and post a narrower one if I get a detailed grep on the problem.

**Comment by sphuber** [comment]
Jens, I think I can give you some piece of mind. I just ran into the exact same issue you described. So you can rest easy now, you are not going crazy :)

I also have the same issue, where I have a workchain calling a subworkchain and the workchain tries to add the results of the subworkchain to its own outputs. If this is done in a step that also uses `self.report` the outputs do not get attached. I will try and find out why so I am reopening this

**Comment by sphuber** [comment]
It seemed like the problem may have had something to do with the entry points and the new plugin system. I noticed that, together with the weird behavior of outputs not being registered, that type strings of classes loaded from plugins randomly seemed to be different and lead to issues. I tried various combinations of cleaning and reinstalling the relevant plugins, as well as calling `reentry scan`, but to no avail. Eventually I tried restarting the daemon, which temporarily has solved the issue. I am not sure what the real problem was, but restarting the daemon was essential.
